pyani_plus/tools.py

```diff
@@ -0,0 +1,197 @@
+typically by executing them and parsing the stdout. These can be used for
+logging, as the exact versions used can sometimes be important for full
+reproducibility of results.
+"""
+
+import re
+import shutil
+import subprocess
+from pathlib import Path
+from typing import NamedTuple
+
+
+class ExternalToolData(NamedTuple):
+    """Convenience struct for tool path and version information."""
+
+    exe_path: Path
+    version: str
+
+
+def check_cmd(cmd: str | Path) -> Path:
+    """Return full Path of given command, or raise RuntimeError."""
+    if not cmd:
+        msg = "Function check_cmd requires a command or full path."
+        raise ValueError(msg)
+
+    exe_str = shutil.which(cmd)
+    if not exe_str:
+        msg = (
+            f"{cmd} is not executable"
+            if "/" in str(cmd)
+            else f"{cmd} not found on $PATH"
+        )
+        raise RuntimeError(msg)
+
+    # Turn into an absolute path in case later use a temp working dir:
+    return Path(exe_str).absolute()
+
+
+def _get_path_and_version_output(
+    cmd: str | Path, args: list[str] | None = None
+) -> tuple[Path, str]:
+    """Determine path of command, run it with args, capture combined stdout and stderr."""
+    # Might later need to add check=True as an optional argument,
+    # e.g. NCBI legacy blast doesn't have a version option and uses return code 1.
+    exe_path = check_cmd(cmd)
+    result = subprocess.run(
+        [str(exe_path), *(args if args else [])],
+        shell=False,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        check=True,
+        text=True,
+    )
+    assert isinstance(result.stdout, str)  # noqa: S101
+    return exe_path, result.stdout
+
+
+def get_blastn(cmd: str | Path = "blastn") -> ExternalToolData:
+    """Return NCBI BLAST+ blastn path and version as a named tuple.
+
+    We expect the tool to behave as follows:
+
+    .. code-block:: bash
+
+        $ blastn -version
+        blastn: 2.16.0+
+         Package: blast 2.16.0, build Aug  6 2024 15:58:44
+
+    Here the function would return the binary path and "2.16.0+" as the version.
+
+    Raises a RuntimeError if the command cannot be found, run, or if the
+    version cannot be inferred - this likely indicates a dramatically
+    different version of the tool with different behaviour.
+    """
+    exe_path, output = _get_path_and_version_output(cmd, ["-version"])
+
+    match = re.search(r"(?<=blastn:\s)[0-9\.]*\+", output)
+    version = match.group().strip() if match else None
+    if not version:
+        msg = f"Executable exists at {exe_path} but could not retrieve version"
+        raise RuntimeError(msg)
+
+    return ExternalToolData(exe_path, version)
+
+
+def get_fastani(cmd: str | Path = "fastani") -> ExternalToolData:
+    """Return FastANI path and version as a named tuple.
+
+    We expect the tool to behave as follows:
+
+    .. code-block:: bash
+
+        $ fastani -v
```

Comment:
Should we make this fastANI -v? fastANI should work on both case-insensitive and case-sensitive filesystems.

Reply:
Yes - I was initially doing this on macOS